File: 프로그래머스/2/42586. 기능개발/기능개발.py
from collections import deque
def solution(progresses, speeds):
    # 각 기능 완성도 progresses
    # 하루 작업 가능량 speeds ex 하루에 1프로씩 작업 가능
    
    # 먼저 있는게 먼저 나간다 -> 큐
    
    N = len(progresses)
    answer = []
    
    progress_queue = deque(progresses)
    speed_queue = deque(speeds)
    
    while progress_queue:
        # 하루 작업 진행
        for i in range(len(progress_queue)):
            progress_queue[i] +=speed_queue[i]
            
        # 배포 가능한 기능 확인
        deploy_count = 0
        while progress_queue and progress_queue[0] >=100:
            progress_queue.popleft()
            speed_queue.popleft()
            deploy_count +=1
            
        # 배포된 기능이 있다면 answer에 추가
        if deploy_count >0 :
            answer.append(deploy_count)
    
    
    
    
    return answer

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;
class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        
        // progresses -> 작업의 진도가 적혀있음(100이 만점)
        // speeds -> 각 인덱스 작업 별 진행 속도
        
        // 각 기능은 앞에 있는 기능이 배포되어야 배포할 수 있다(먼저 100이되어서 배포 불가)
        
        // 각 배포마다 몇개의 기능이 배포되는지
        List<Integer> result = new ArrayList<>();
        
        int n = progresses.length; 
        
        // 각 업무가 얼마나 걸릴려나
        int[] days = new int[n];
        
        for(int i = 0; i < n; i++) {
            int remain = 100 - progresses[i];
            int day = remain / speeds[i];
            if(remain % speeds[i] != 0) {
                day++;
            }
        
            days[i] = day;
        }
        
        // 지금 일이 끝나는 시점에 뒤에 일이 끝나 있나?
        
        for(int i  = 0 ;  i < n ;) {
            // 지금 일이 끝나는 일수
            int cur = days[i];
            int count = 1;
            
            int j = i + 1;
            while (j < n && days[j] <= cur) {
                count++;
                j++;
            }
            
            i= j;

            result.add(count);
        }
        
        int answerSize = result.size();
        int[] answer = new int[answerSize];
        
        for(int i=0; i<answerSize;i++){
            answer[i] = result.get(i);
        }
        
        return answer;
    }
}